src/nxt_cgroup.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+
+#include <nxt_cgroup.h>
+
+
+static int nxt_mk_cgpath_relative(nxt_task_t *task, const char *dir,
+    char *cgpath);
+static nxt_int_t nxt_mk_cgpath(nxt_task_t *task, const char *dir,
+    char *cgpath);
+
+
+nxt_int_t
+nxt_cgroup_proc_add(nxt_task_t *task, nxt_process_t *process)
+{
+    int        len;
+    char       cgprocs[NXT_MAX_PATH_LEN];
+    FILE       *fp;
+    nxt_int_t  ret;
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.path == NULL)
+    {
+        return NXT_OK;
+    }
+
+    ret = nxt_mk_cgpath(task, process->isolation.cgroup.path, cgprocs);
+    if (nxt_slow_path(ret == NXT_ERROR)) {
+        return NXT_ERROR;
+    }
+
+    ret = nxt_fs_mkdir_all((const u_char *) cgprocs, 0777);
+    if (nxt_slow_path(ret == NXT_ERROR)) {
+        return NXT_ERROR;
+    }
+
+    len = strlen(cgprocs);
+
+    len = snprintf(cgprocs + len, NXT_MAX_PATH_LEN - len, "/cgroup.procs");
+    if (nxt_slow_path(len >= NXT_MAX_PATH_LEN)) {
```

Comment:
Oh, I didn't realize before, but the bug I mentioned in private still seems to be there:

`len` must be checked against `PATH_MAX - len`, not just `PATH_MAX`, since that's what snprintf(3) got.

However, the underlying issue here is using an API that was not designed for that.  We should either use snprintf(3) as it's been designed (with its performance issues; [Shliemel the painer](https://statmodeling.stat.columbia.edu/2013/11/04/joel-on-shlemiel-the-painter-unknown-unknowns/); yada yada).

In fact, you were trying to use a home-made [`strscpy(9)`](https://manpages.debian.org/testing/linux-manual-4.8/strscpy.9.en.html).  Still, strscpy(9) has issues, and I wouldn't want to use it here either.

I prefer my own `stpecpy()`, as mentioned earlier, which would be used as (and assuming the `nxt_mk_cgpath()` call has been improved to return a pointer to the end:

```c
    past_end = cgprocs + sizeof(cgprocs);

    p = nxt_mk_cgpath(task, process->isolation.cgroup.path, cgprocs);
    if (nxt_slow_path(p == NULL)) {
        return NXT_ERROR;
    }

    ret = nxt_fs_mkdir_all((const u_char *) cgprocs, 0777);
    if (nxt_slow_path(ret == NXT_ERROR)) {
        return NXT_ERROR;
    }

    p = snprintf(p, past_end, "/cgroup.procs");
    if (nxt_slow_path(p == past_end)) {
        nxt_errno = ENAMETOOLONG;
        return NXT_ERROR;
    }
```

Reply:
```c
char *
stpecpy(char *dst, const char *restrict src, char *past_end)
{
	char *p;

	if (dst == past_end)
		return past_end;

	p = memccpy(dst, src, '\0', past_end - dst);
	if (p != NULL)
		return p - 1;

	/* truncation detected */
	past_end[-1] = '\0';
	return past_end;
}
```

Reply:
No. I was definitely using snprintf(), not an uncommon use of it...

Reply:
I've updated the check.

If we ever introduce yet another string API, we can switch to it, but for now we've got what we've got.

Reply:
> No. I was definitely using snprintf(), not an uncommon use of it...

Yes it's a common use.  But I'd call it just a common workaround for not having a good API for chain-copying, or really anything better at all in libc.

So, the patches look good to me (in the sense that I see no bugs), but they look quite ugly to me in the sense that they're workarounding the absence of a good API that allows to "chain-copy a string into another string with truncation".  Please remove the Reviewed-by tag from this patch (but the other patches are fine; and in fact you can also add it for the nxt_fopen/nxt_fclose patch).  While I won't veto it, I certainly don't want to sign it.

Reply:
I'll propose an improvement after we apply this patch set.

Reply:
> Please remove my Reviewed-by tag from this patch (but the other patches are fine; and in fact you can also add it for the nxt_fopen/nxt_fclose patch). While I won't veto it, I certainly don't want to sign it.

Replacing it by `Cc:` would make sense.

Reply:
As you wish...